src/apps/config/alarm_codec.lua

```diff
@@ -0,0 +1,195 @@
+-- Use of this source code is governed by the Apache 2.0 license; see COPYING.
+
+module(...,package.seeall)
+
+local S = require("syscall")
+local channel = require("apps.config.channel")
+local ffi = require("ffi")
+
+local alarm_names = { 'raise_alarm', 'clear_alarm' }
+local alarm_codes = {}
+for i, name in ipairs(alarm_names) do alarm_codes[name] = i end
+
+local alarms = {}
+
+function alarms.raise_alarm (codec, resource, alarm_type_id, alarm_type_qualifier,
+   perceived_severity, alarm_text)
+
+   local resource = codec:string(resource)
+   local alarm_type_id = codec:string(alarm_type_id)
+   local alarm_type_qualifier = codec:string(alarm_type_qualifier)
+
+   local perceived_severity = codec:string(perceived_severity)
+   local alarm_text = codec:string(alarm_text)
+
+   return codec:finish(resource, alarm_type_id, alarm_type_qualifier,
+                       perceived_severity, alarm_text)
+end
+function alarms.clear_alarm (codec, resource, alarm_type_id, alarm_type_qualifier)
+   local resource = codec:string(resource)
+   local alarm_type_id = codec:string(alarm_type_id)
+   local alarm_type_qualifier = codec:string(alarm_type_qualifier)
+
+   return codec:finish(resource, alarm_type_id, alarm_type_qualifier)
+end
+
+local function encoder()
+   local encoder = { out = {} }
+   function encoder:uint32(len)
+      table.insert(self.out, ffi.new('uint32_t[1]', len))
+   end
+   function encoder:string(str)
+      self:uint32(#str)
+      local buf = ffi.new('uint8_t[?]', #str)
+      ffi.copy(buf, str, #str)
+      table.insert(self.out, buf)
+   end
+   function encoder:finish()
+      local size = 0
+      for _,src in ipairs(self.out) do size = size + ffi.sizeof(src) end
+      local dst = ffi.new('uint8_t[?]', size)
+      local pos = 0
+      for _,src in ipairs(self.out) do
+         ffi.copy(dst + pos, src, ffi.sizeof(src))
+         pos = pos + ffi.sizeof(src)
+      end
+      return dst, size
+   end
+   return encoder
+end
+
+function encode(alarm)
```

Comment:
Likewise here I think it's probably OK to directly expose e.g. `encode_raise_alarm` instead of bundling outgoing alarms through a table.  For a followup.